### PR TITLE
separate download code to two parts - close via defer

### PR DESCRIPTION
**Pull request by JaSei**
This should be cover all download error and valid close filehandle

### CODE REVIEW

File: client/download.go [hunk 1]
```diff
@@ -138,6 +138,21 @@ func downloadFile(httpClient httpClient, filepath string, url string, devnull bo
 		}
 	}
 
+	size, err = downloadFile(httpClient, temppath, url, devnull, expectedSha)
+	if err != nil {
+		return size, err
+	}
+
+	if !devnull {
```
Comment (augi) [inline client/download.go]: If `devnull` is `true` then we should delete the file IMHO.

**Review by augi — CHANGES_REQUESTED** [review]

File: client/download.go (hunk 1 above)
Comment (JaSei) [reply client/download.go]: good point...
`devnull` really complicate this code :wink:

Comment (augi) [reply client/download.go]: Is it even needed to have `devnull` functionality? Wouldn't be better to remove it?

File: client/download.go [hunk 2]
```diff
@@ -138,6 +147,35 @@ func downloadFile(httpClient httpClient, filepath string, url string, devnull bo
 		}
 	}
 
+	size, err = downloadFile(httpClient, temppath, url, expectedSha)
+	if err != nil {
+		return size, err
```
Comment (augi) [inline client/download.go]: We should remove the tempfile here IMHO.

**Review by augi — CHANGES_REQUESTED** [review]

File: client/download.go (hunk 2 above)
Comment (JaSei) [reply client/download.go]: right, thanks

**Review by augi — APPROVED** [review]

**Pull request merged**